zerver/lib/message.py

```diff
@@ -1557,73 +1557,63 @@ def get_recent_conversations_recipient_id(
+    It may be possible to write this query directly in Django, however
+    it is made much easier by using CTEs, which Django does not
+    natively support.
 
     We return a dictionary structure for convenient modification
     below; this structure is converted into its final form by
     post_process.
 
     """
-    RECENT_CONVERSATIONS_LIMIT = 1000
+    RECENT_CONVERSATIONS_LIMIT = 10000
 
     recipient_map = {}
     my_recipient_id = user_profile.recipient_id
 
     query = SQL(
         """
-    SELECT
-        subquery.recipient_id, MAX(subquery.message_id)
-    FROM (
-        (SELECT
-            um.message_id AS message_id,
-            m.recipient_id AS recipient_id
-        FROM
-            zerver_usermessage um
-        JOIN
-            zerver_message m
-        ON
-            um.message_id = m.id
-        WHERE
-            um.user_profile_id=%(user_profile_id)s AND
-            um.flags & 2048 <> 0 AND
-            m.recipient_id <> %(my_recipient_id)s
-        ORDER BY message_id DESC
-        LIMIT %(conversation_limit)s)
-        UNION ALL
-        (SELECT
-            m.id AS message_id,
-            sender_profile.recipient_id AS recipient_id
-        FROM
-            zerver_message m
-        JOIN
-            zerver_userprofile sender_profile
-        ON
-            m.sender_id = sender_profile.id
-        WHERE
-            m.realm_id=%(realm_id)s AND
-            m.recipient_id=%(my_recipient_id)s
-        ORDER BY message_id DESC
-        LIMIT %(conversation_limit)s)
-    ) AS subquery
-    GROUP BY subquery.recipient_id
+        WITH personals AS (
+            SELECT   um.message_id AS message_id
+            FROM     zerver_usermessage um
+            WHERE    um.user_profile_id = %(user_profile_id)s
+            AND      um.flags & 2048 <> 0
+            ORDER BY message_id DESC limit %(conversation_limit)s
+        ),
+        message AS (
+            SELECT message_id,
+                   CASE
+                          WHEN m.recipient_id = %(my_recipient_id)s
+                          THEN m.sender_id
+                          ELSE NULL
+                   END AS sender_id,
+                   CASE
+                          WHEN m.recipient_id <> %(my_recipient_id)s
+                          THEN m.recipient_id
+                          ELSE NULL
+                   END AS outgoing_recipient_id
+            FROM   personals
+            JOIN   zerver_message m
+            ON     personals.message_id = m.id
+        ),
+        unified AS (
+            SELECT    message_id,
+                      COALESCE(zerver_userprofile.recipient_id, outgoing_recipient_id) AS other_recipient_id
+            FROM      message
+            LEFT JOIN zerver_userprofile
+            ON        zerver_userprofile.id = sender_id
+        )
+        SELECT   other_recipient_id,
+                 MAX(message_id)
+        FROM     unified
+        GROUP BY other_recipient_id
```

Comment:
We should make sure we have tests for the DMs to yourself on this code path, since that's the kind of thing one could imagine going awry with a change to this query (I don't think it will).